tests/cupy_tests/cuda_tests/test_cub.py

```diff
@@ -0,0 +1,159 @@
+                    impl(self, *args, **kw)
+                except Exception:
+                    print(name, 'is', a, ', ndim is', ndim, ', shape is',
+                          self.shape)
+                    raise
+        return test_func
+    return decorator
+
+
+def timing(test, func, arr, axis, runs=20):
+    cupy.cuda.cub_enabled = True
+    t_cub = 0.
+    for i in range(runs):
+        test.start.record()
+        getattr(arr, func)(axis=axis)
+        test.stop.record()
+        test.stop.synchronize()
+        t_cub += cupy.cuda.get_elapsed_time(test.start, test.stop)
+
+    cupy.cuda.cub_enabled = False
+    t_cupy = 0.
+    for i in range(runs):
+        test.start.record()
+        getattr(arr, func)(axis=axis)
+        test.stop.record()
+        test.stop.synchronize()
+        t_cupy += cupy.cuda.get_elapsed_time(test.start, test.stop)
+
+    # TODO(leofang): raise PerformanceWarning here?
+    print("CUB: {:10.5f}; CuPy: {:10.5f} (ms), for {} runs, shape={}, axis={},"
+          " dtype={}".format(t_cub, t_cupy, runs, arr.shape, axis, arr.dtype))
+    cupy.cuda.cub_enabled = True  # restore
+
+
+# This class compares CUB results against NumPy's
+@testing.parameterize(*testing.product({
+    'shape': [(10,), (10, 20), (10, 20, 30), (10, 20, 30, 40)],
+}))
+@testing.gpu
+class TestCUBreduction(unittest.TestCase):
+    @for_contiguous_axes()
+    # sum supports less dtypes; don't test float16 as it's not as accurate?
+    @testing.for_dtypes('lLfdFD')
+    @testing.numpy_cupy_allclose(rtol=1E-5)
+    def test_cub_sum(self, xp, dtype, axis):
+        assert cupy.cuda.cub_enabled
+        a = testing.shaped_random(self.shape, xp, dtype)
+        return a.sum(axis=axis)
+
+    @for_contiguous_axes()
+    @testing.for_dtypes('bhilBHILefdFD')
+    @testing.numpy_cupy_allclose(rtol=1E-5)
+    def test_cub_min(self, xp, dtype, axis):
+        assert cupy.cuda.cub_enabled
+        a = testing.shaped_random(self.shape, xp, dtype)
+        return a.min(axis=axis)
+
+    @for_contiguous_axes()
+    @testing.for_dtypes('bhilBHILefdFD')
+    @testing.numpy_cupy_allclose(rtol=1E-5)
+    def test_cub_max(self, xp, dtype, axis):
+        assert cupy.cuda.cub_enabled
+        a = testing.shaped_random(self.shape, xp, dtype)
+        return a.max(axis=axis)
+
+    # argmin does not support axis yet
+    @testing.for_dtypes('bhilBHILefdFD')
+    @testing.numpy_cupy_allclose(rtol=1E-5)
+    def test_cub_argmin(self, xp, dtype):
+        assert cupy.cuda.cub_enabled
+        a = testing.shaped_random(self.shape, xp, dtype)
+        return a.argmin()
+
+    # argmax does not support axis yet
+    @testing.for_dtypes('bhilBHILefdFD')
+    @testing.numpy_cupy_allclose(rtol=1E-5)
+    def test_cub_argmax(self, xp, dtype):
+        assert cupy.cuda.cub_enabled
+        a = testing.shaped_random(self.shape, xp, dtype)
+        return a.argmax()
+
+
+# This class compares CUB results against CuPy's original implementation
```

Comment:
Right, I noticed that too. This part was meant to detect performance regression. CUB is generally faster, but there could be exceptions. Do you think we don’t need to worry about it for now?

Reply:
Let me clarify, what do you mean by "performance regression"? Do you want to find which parameter combinations are slower when CUB is enabled (ie, find "exceptions") ? or you want to detect regression between versions/commits?

Reply:
This class has the former (finding exceptions) in mind. It compares with the old reduction kernels. 

I don’t think the latter is possible without having some persistent database recording the performance data. It’s not easy to set this up...

Reply:
I removed the performance tests in the refactored tests. Let me know if you want me to bring them back in an alternative form. Presumably it could also be another PR.